### PR TITLE
engine: Update API for C++17

### PR DESCRIPTION
**Pull request by mwestphal**
- Rewrite getPluginsList
 - Update API for C++17
 - Add/Update test for better testing and coverage
 - Improve some exceptions doc and logic

### CODE REVIEW

File: library/src/engine.cxx [hunk 1]
```diff
@@ -41,16 +41,16 @@ engine::engine(
 
   // build default cache path
 #if defined(_WIN32)
-  std::string cachePath = vtksys::SystemTools::GetEnv("LOCALAPPDATA");
-  cachePath = cachePath + "/f3d";
+  fs::path cachePath(std::getenv("LOCALAPPDATA"));
```
Comment (Meakk) [inline library/src/engine.cxx:44]: Creating a `fs::path` with nullptr will crash the app. You need to check the result of `std::getenv`

File: library/src/engine.cxx [hunk 2]
```diff
@@ -213,7 +213,7 @@ interactor& engine::getInteractor()
 }
 
 //----------------------------------------------------------------------------
-void engine::loadPlugin(const std::string& pathOrName, const std::vector<std::string>& searchPaths)
+void engine::loadPlugin(std::string_view pathOrName, const std::vector<fs::path>& searchPaths)
```
Comment (Meakk) [inline library/src/engine.cxx:216]: `pathOrName` is used many times to create a string, I think keeping it const ref might be better?

File: library/src/engine.cxx [hunk 3]
```diff
@@ -364,6 +364,9 @@ std::vector<std::string> engine::getPluginsList(const std::string& pluginPath)
       }
     }
   }
+  catch (const std::filesystem::filesystem_error&)
```
Comment (Meakk) [inline library/src/engine.cxx:367]:
```suggestion
  catch (const fs::filesystem_error&)
```

File: library/src/engine.cxx [hunk 4]
```diff
@@ -442,7 +445,7 @@ std::vector<engine::readerInformation> engine::getReadersInfo()
 }
 
 //----------------------------------------------------------------------------
-engine& engine::setCachePath(const std::string& cachePath)
+engine& engine::setCachePath(const std::filesystem::path& cachePath)
```
Comment (Meakk) [inline library/src/engine.cxx:448]:
```suggestion
engine& engine::setCachePath(const fs::path& cachePath)
```

File: library/src/window_impl.cxx [hunk 5]
```diff
@@ -549,9 +549,9 @@ void window_impl::SetImporter(vtkF3DMetaImporter* importer)
 }
 
 //----------------------------------------------------------------------------
-void window_impl::SetCachePath(const std::string& cachePath)
+void window_impl::SetCachePath(const std::filesystem::path& cachePath)
 {
-  this->Internals->CachePath = cachePath;
+  this->Internals->CachePath = cachePath.string();
```
Comment (Meakk) [inline library/src/window_impl.cxx:554]: Maybe `CachePath` should be a fs::path too?

File: library/testing/TestSDKEngine.cxx [hunk 6]
```diff
@@ -71,5 +72,9 @@ int TestSDKEngine(int argc, char* argv[])
     return EXIT_FAILURE;
   }
 
+  // coverage for getPluginList
+  std::ignore = f3d::engine::getPluginsList(std::string(argv[1]) + "configs");
+  std::ignore = f3d::engine::getPluginsList("innexistent");
```
Comment (Meakk) [inline library/testing/TestSDKEngine.cxx:77]:
```suggestion
  std::ignore = f3d::engine::getPluginsList("inexistent");
```

File: library/testing/TestSDKEngineExceptions.cxx [hunk 7]
```diff
@@ -1,99 +1,82 @@
...
+      "olderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfol"
+      "derfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolde"
+      "rfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderf"
+      "olderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfol"
+      "derfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolde"
+      "rfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderf"
+      "olderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfol"
+      "derfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolde"
+      "rfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderf"
+      "olderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolderfolder/file.ext");
```
Comment (Meakk) [inline library/testing/TestSDKEngineExceptions.cxx:78]: Very verbose, what about this?
```suggestion
    f3d::engine::loadPlugin("/" + std::string(1000, 'x') + "/file.ext");
```

**Review by Meakk — CHANGES_REQUESTED** [review]

File: application/F3DPluginsTools.cxx [hunk 8]
```diff
@@ -6,21 +6,27 @@
...
 {
 //----------------------------------------------------------------------------
-std::vector<std::string> GetPluginSearchPaths()
+std::vector<fs::path> GetPluginSearchPaths()
 {
+  std::vector<fs::path> searchPaths;
+
   // Recover F3D_PLUGINS_PATH first
-  auto searchPaths = F3DSystemTools::GetVectorEnvironnementVariable("F3D_PLUGINS_PATH");
+  std::vector<std::string> stringPaths =
```
[the remark this answers is not in the document]
Comment (mwestphal) [reply application/F3DPluginsTools.cxx:19]: definitely not, this function can be used for any env var, it is not path specific.

File: library/src/engine.cxx (hunk 1 above)
Comment (mwestphal) [reply library/src/engine.cxx:44]: good catch!

File: library/src/window_impl.cxx (hunk 5 above)
Comment (mwestphal) [reply library/src/window_impl.cxx:554]: indeed

File: library/src/engine.cxx [hunk 9]
```diff
@@ -39,18 +39,29 @@ engine::engine(
...
+
+#if defined(_WIN32)
+  cachePath /= "f3d";
 #elif defined(__APPLE__)
-  std::string cachePath = vtksys::SystemTools::GetEnv("HOME");
-  cachePath += "/Library/Caches/f3d";
+  cachePath = cachePath / "Library" / "Caches" / "f3d";
 #elif defined(__ANDROID__)
-  std::string cachePath = ""; // no default
+//  fs::path cachePath;
```
Comment (mwestphal) [inline library/src/engine.cxx:62]: no default does not makes much sense, we cannot force libf3d user to set a cachePath. Is there no reasonnable cache path we could use by default ?

Comment (mwestphal) [reply library/src/engine.cxx:62]: @Meakk

File: application/F3DPluginsTools.cxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Meakk) [reply application/F3DPluginsTools.cxx:19]: I meant the `GetPluginSearchPaths()` function but it seems like you changed it.

File: library/src/window_impl.cxx [hunk 10]
```diff
@@ -128,38 +121,39 @@ window_impl::window_impl(const options& options, const std::optional<Type>& type
   {
 #if defined(VTK_OPENGL_HAS_EGL)
     this->Internals->RenWin = vtkSmartPointer<vtkF3DEGLRenderWindow>::New();
-#else
-    throw engine::no_window_exception("Window type is EGL but VTK EGL support is not enabled");
 #endif
+    // No #else needed, it would be unreachable
```
Comment (Meakk) [inline library/src/window_impl.cxx:125]: Why isn't it unreachable? `VTK_OPENGL_HAS_EGL` can be OFF

Comment (Meakk) [reply library/src/window_impl.cxx:125]: Good idea, I think `#error` will help us in the future if we change something in `context` class.

Comment (mwestphal) [reply library/src/window_impl.cxx]: obv #error wont work as this is still compiled, I've added asserts

**Review by Meakk — APPROVED** [review]

**Pull request merged**